File: LeetCodeRecord/ACFirst/230-KthSmallest.java
package ACFirst;

class KthSmallest {
        
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public int res = 0;
    public int k = 0;

    // 中序遍历的框架
    public int kthSmallest(TreeNode root, int k) {
        
        this.k = k;
        dfs(root);

        return res;
    }

    public void dfs(TreeNode root) {
        if (root == null) {
            return;
        }

        dfs(root.left);

        if (k == 0) {
            return;
        }

        if (--k == 0) {
            res = root.val;
            return;
        }

        dfs(root.right);
    }
}

File: Top200/230-KthSmallest.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func kthSmallest(root *TreeNode, k int) int {
	var res int = 0
	dfs(root, &k, &res)

	return res
}

func dfs(root *TreeNode, k, res *int) {
	if root == nil {
		return
	}

	dfs(root.Left, k, res)

	if *k == 0 {
		return
	}

	*k = *k - 1
	if *k == 0 {
		*res = root.Val
		return
	}

	dfs(root.Right, k, res)
}
